exasol_integration_test_docker_environment/lib/base/db_os_executor.py

```diff
@@ -0,0 +1,100 @@
+from abc import abstractmethod
+import fabric
+import docker
+from typing import Protocol, runtime_checkable
+from docker.models.containers import Container, ExecResult
+from exasol_integration_test_docker_environment \
+    .lib.base.ssh_access import SshKey
+from exasol_integration_test_docker_environment \
+    .lib.data.database_info import DatabaseInfo
+
+
+@runtime_checkable
+class DbOsExecutor(Protocol):
+    """
+    This class provides an abstraction to execute commands inside a Docker
+    Container.  See concrete implementations in sub-classes
+    ``PlainDockerExec`` and ``SshDockerExec``.
+    """
+    @abstractmethod
+    def exec(self, cmd: str) -> ExecResult:
+        ...
+
+
+class DockerExecutor(DbOsExecutor):
+    def __init__(self, container_name: str, timeout: int = 100000):
+        self.container_name = container_name
+        self.timeout = timeout
+        self.container = None
+
+    def _get_docker_client(self):
+        return ContextDockerClient(timeout=self.timeout)
+
+    def _find_container():
+        if not self.container:
+            with self._get_docker_client() as client:
```

Comment:
I can do so, but maybe it would also be worthwhile to talk a little about the life cycle of the instances of `DockerFactory` and the sub classes of `DbOsExecutor`. Maybe in some places we should pass the executor rather than the factory?

Reply:
Lifetime of an executor is one luigi task.